File: MCPForUnity/Editor/Tools/RerunFailedTests.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using MCPForUnity.Editor.Helpers;
using MCPForUnity.Editor.Services;
using Newtonsoft.Json.Linq;

namespace MCPForUnity.Editor.Tools
{
    /// <summary>
    /// Reruns the failed tests from a previous Unity test run.
    /// When no run id is provided, the most recent completed run is used.
    /// </summary>
    [McpForUnityTool("rerun_failed_tests")]
    public static class RerunFailedTests
    {
        private const int DefaultTimeoutSeconds = 600;

        public static async Task<object> HandleCommand(JObject @params)
        {
            string runId = @params?["runId"]?.ToString();
            bool waitForCompletion = ParseBool(@params?["waitForCompletion"], defaultValue: true);
            int timeoutSeconds = ParseTimeout(@params, DefaultTimeoutSeconds);

            var testService = MCPServiceLocator.Tests;
            IReadOnlyList<TestRunTestResult> failed = testService.GetFailedTests(runId);
            if (failed == null || failed.Count == 0)
            {
                string message = string.IsNullOrEmpty(runId)
                    ? "No failed tests available from the most recent completed run."
                    : $"No failed tests recorded for run '{runId}'.";
                return Response.Error(message);
            }

            if (!testService.TryGetStatus(runId, out var baseStatus) || baseStatus == null)
            {
                string message = string.IsNullOrEmpty(runId)
                    ? "Unable to locate the last completed run to rerun."
                    : $"Test run '{runId}' not found.";
                return Response.Error(message);
            }

            var failedNames = failed
                .Select(f => f.FullName)
                .Where(n => !string.IsNullOrWhiteSpace(n))
                .Distinct(StringComparer.Ordinal)
                .ToList();

            if (failedNames.Count == 0)
            {
                return Response.Error("Failed tests do not contain valid identifiers to rerun.");
            }

            var request = new TestRunRequest(baseStatus.Mode, failedNames);

            TestRunHandle handle;
            try
            {
                handle = testService.StartRun(request);
            }
            catch (Exception ex)
            {
                return Response.Error($"Failed to start rerun: {ex.Message}");
            }

            if (!waitForCompletion)
            {
                testService.TryGetStatus(handle.RunId, out var snapshot);
                var payload = snapshot?.ToSerializable() ?? new { runId = handle.RunId, state = TestRunState.Queued.ToString() };
                string message = handle.StartedNewRun
                    ? $"Started rerun '{handle.RunId}' for failed tests."
                    : $"Reusing active run '{handle.RunId}'.";
                return Response.Success(message, payload);
            }

            var timeoutTask = Task.Delay(TimeSpan.FromSeconds(timeoutSeconds));
            var completed = await Task.WhenAny(handle.CompletionTask, timeoutTask).ConfigureAwait(true);
            if (completed != handle.CompletionTask)
            {
                testService.TryGetStatus(handle.RunId, out var snapshot);
                var data = snapshot?.ToSerializable() ?? new { runId = handle.RunId, state = TestRunState.Running.ToString() };
                return Response.Error($"Failed-test rerun '{handle.RunId}' is still running after {timeoutSeconds} seconds.", data);
            }

            var result = await handle.CompletionTask.ConfigureAwait(true);
            string summary =
                $"Rerun '{result.RunId}' finished with state {result.Summary.ResultState}: {result.Passed}/{result.Total} passed, {result.Failed} failed, {result.Skipped} skipped.";
            return Response.Success(summary, result.ToSerializable());
        }

        private static int ParseTimeout(JObject @params, int fallback)
        {
            try
            {
                var timeoutToken = @params?["timeoutSeconds"];
                if (timeoutToken != null && int.TryParse(timeoutToken.ToString(), out var parsed) && parsed > 0)
                {
                    return parsed;
                }
            }
            catch
            {
                // Ignore and use fallback
            }

            return fallback;
        }

        private static bool ParseBool(JToken token, bool defaultValue)
        {
            if (token == null)
            {
                return defaultValue;
            }

            if (bool.TryParse(token.ToString(), out var parsed))
            {
                return parsed;
            }

            return defaultValue;
        }
    }
}

File: Server/tools/rerun_failed_tests.py
"""Tool for rerunning only the failed Unity tests from a prior run."""
from typing import Annotated, Any

from fastmcp import Context
from pydantic import Field

from registry import mcp_for_unity_tool
from tools import async_send_with_unity_instance, get_unity_instance_from_context
from unity_connection import async_send_command_with_retry

from .run_tests import (
    DefaultTimeoutSeconds,
    RunCompletionTimeout,
    RunTestsResponse,
    build_summary_message,
    coerce_bool,
    coerce_int,
    wait_for_run_completion,
)


@mcp_for_unity_tool(description="Reruns failed Unity tests from a previous run")
async def rerun_failed_tests(
    ctx: Context,
    run_id: Annotated[str | None, Field(
        description="Optional test run identifier to pull failures from; defaults to the last completed run"
    )] = None,
    wait_for_completion: Annotated[bool | str | None, Field(
        description="Set false to return immediately after starting the rerun"
    )] = True,
    timeout_seconds: Annotated[int | str | None, Field(
        description="Optional timeout in seconds for the rerun (string, e.g. '30')"
    )] = None,
) -> RunTestsResponse:
    unity_instance = get_unity_instance_from_context(ctx)

    params: dict[str, Any] = {}
    if run_id:
        params["runId"] = run_id

    wait = coerce_bool(wait_for_completion, default=True)
    params["waitForCompletion"] = False

    timeout = coerce_int(timeout_seconds)
    wait_timeout = timeout or DefaultTimeoutSeconds
    if timeout is not None:
        params["timeoutSeconds"] = timeout

    response = await async_send_with_unity_instance(
        async_send_command_with_retry,
        unity_instance,
        "rerun_failed_tests",
        params,
    )
    if not isinstance(response, dict):
        return response

    if not wait:
        return RunTestsResponse(**response)

    run_id_from_response = (response.get("data") or {}).get("runId") or run_id
    if not run_id_from_response:
        return RunTestsResponse(**response)

    try:
        final_data = await wait_for_run_completion(ctx, unity_instance, run_id_from_response, wait_timeout)
    except RunCompletionTimeout as exc:
        snapshot = exc.snapshot or {"runId": run_id_from_response, "state": "Unknown"}
        message = f"Failed-test rerun '{run_id_from_response}' is still running after {exc.timeout_seconds} seconds."
        return RunTestsResponse(success=False, message=message, error=message, data=snapshot)
    except Exception as exc:
        error_message = f"Failed to retrieve rerun results for '{run_id_from_response}': {exc}"
        return RunTestsResponse(success=False, message=error_message, error=str(exc), data={"runId": run_id_from_response})

    summary_message = build_summary_message(run_id_from_response, final_data)
    return RunTestsResponse(success=True, message=summary_message, data=final_data)
